### ISSUE TITLE
Question regarding use of CollectionStorage.AsArray

### ISSUE THREAD

**Issue by mronnblom-cint** [JasperFx/marten#1386]
Hi!

I'm trying to get the Serializer option "CollectionStorage" to work, but when I set it to "AsArray", I only get null in the database. The current implementation is a IReadOnlyCollection&lt;int> wrapping a List&lt;int>, but I have tried with several others and none of them work:

- IReadOnlyCollection&lt;int> wrapping List&lt;int>
- IEnumerable&lt;int> wrapping List&lt;int>
- IEnumerable&lt;int> wrapping int[]

```
public class Foo
{
	public IReadOnlyCollection<int> Categories { get; private set; }

	public Foo(IEnumerable<int> categories)
	{
		Categories = categories.ToList();
	}
}

```

```
var documentStore = DocumentStore.For(_ =>
{
	_.Serializer(new JsonNetSerializer
	{
		CollectionStorage = CollectionStorage.AsArray
	});
	_.Schema.For<Foo>();
        ...
}
```

It only works if the property type is an array, but I don't want the property to be changeable from outside the object (DDD). I.e. this works:
```
public class Foo
{
	public int[] Categories { get; private set; }

	public Foo(IEnumerable<int> categories)
	{
		Categories = categories.ToArray();
	}
}

```

The reason for me wanting to set CollectionStorage.AsArray is that collections get rather unwieldy in the db otherwise:
```
"Categories": {
     "$type": "System.Collections.Generic.List`1[[System.Int32, System.Private.CoreLib]], System.Private.CoreLib",
     "$values": [
         1,
         3
     ]
}
```

I'm using Marten 3.9.0, PostgreSQL for Windows v12, .Net Core 3.0.

**Comment by mysticmind** [comment]
@mronnblom-cint I am trying to verify this use case, will get back to you.

**Comment by mysticmind** [comment]
@mronnblom-cint if you use 3.10.0, you can pass `NonPublicMembersStorage = NonPublicMembersStorage.NonPublicSetters` for the serializer and your use case work as expected. Refer to docs [here](https://jasperfx.github.io/marten/documentation/documents/json/newtonsoft/)

**Comment by mronnblom-cint** [comment]
@mysticmind I tried `NonPublicMembersStorage.NonPublicSetters`, but the result is the same - null in db. I don't think that the problem is that the setter is private. I have lots of other properties in the same class that have private setters, and they get written to the db as expected. It is only collection properties that are problematic.

The config I just tried:

```
var documentStore = DocumentStore.For(_ =>
{
	_.Connection(configuration.GetConnectionString(connectionStringKey));
	_.HiloSequenceDefaults.MaxLo = 1;
	_.AutoCreateSchemaObjects = AutoCreate.None;
	_.UseDefaultSerialization(
		EnumStorage.AsString,
		Casing.CamelCase,
		CollectionStorage.AsArray,
		NonPublicMembersStorage.NonPublicSetters);
	_.PLV8Enabled = false;
	_.Schema.For<Foo>()
		.UseOptimisticConcurrency(true);
});
```

**Comment by mronnblom-cint** [comment]
@mysticmind I managed to get the desired behavior if I change the autoproperty to an explicit backing field, like this:

```
public class Foo
{
    // Doesn't work
    //public IReadOnlyCollection<int> Categories { get; private set; }

    // Works
    private List<int> _categories;
    public IReadOnlyCollection<int> Categories => _categories.AsReadOnly();

    public Foo(IEnumerable<int> categories)
    {
        _categories = categories.ToList();
    }
}
```

For some reason, this serializes to an array in the db even _without_ CollectionStorage.AsArray. This is good enough for me, so you may close this issue.

**Comment by oskardudycz** [comment]
@mronnblom-cint thank you for the feedback 👍 I'll double check that, as the `public IReadOnlyCollection<int> Categories { get; private set; }` should also work.

**Comment by oskardudycz** [comment]
@mronnblom-cint the issue was with the combination of `CollectionStorage.AsArray` and `IReadonlyCollection<int>` - for the `IReadonlyCollection<string>` it was working fine. We had only coverage for the string case.

I provided a fix with the additional tests in https://github.com/JasperFx/marten/pull/1390.

Fix should be available in `3.11`.

**Comment by mronnblom-cint** [comment]
Great news! Thanks!
